### PR TITLE
Hooks accept arbitrary bytes from swapper + hook.getFee() accepts parameters

### PR DESCRIPTION
**Pull request by ewilz**
## Related Issue
Which issue does this pull request resolve?
fixes: https://github.com/Uniswap/v4-core/issues/307

about an 850 gas overhead for zero bytes for arbitrary data

## Description of changes
add arbitrary bytes for hooks

### CODE REVIEW

File: contracts/interfaces/IHooks.sol [hunk 1]
```diff
@@ -15,15 +15,17 @@ interface IHooks {
     /// @param key The key for the pool being initialized
     /// @param sqrtPriceX96 The sqrt(price) of the pool as a Q64.96
     /// @return bytes4 The function selector for the hook
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:17]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 2]
```diff
@@ -15,15 +15,17 @@ interface IHooks {
...
+    function beforeInitialize(address sender, PoolKey calldata key, uint160 sqrtPriceX96, bytes memory data)
+        external
+        returns (bytes4);
 
     /// @notice The hook called after the state of a pool is initialized
     /// @param sender The initial msg.sender for the initialize call
     /// @param key The key for the pool being initialized
     /// @param sqrtPriceX96 The sqrt(price) of the pool as a Q64.96
     /// @param tick The current tick after the state of a pool is initialized
     /// @return bytes4 The function selector for the hook
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:27]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 3]
```diff
@@ -47,17 +50,21 @@ interface IHooks {
         address sender,
         PoolKey calldata key,
         IPoolManager.ModifyPositionParams calldata params,
-        BalanceDelta delta
+        BalanceDelta delta,
+        bytes memory data
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:54]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 4]
```diff
@@ -47,17 +50,21 @@ interface IHooks {
...
     ) external returns (bytes4);
 
     /// @notice The hook called before a swap
     /// @param sender The initial msg.sender for the swap call
     /// @param key The key for the pool
     /// @param params The parameters for the swap
     /// @return bytes4 The function selector for the hook
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:61]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 5]
```diff
@@ -69,7 +76,8 @@ interface IHooks {
         address sender,
         PoolKey calldata key,
         IPoolManager.SwapParams calldata params,
-        BalanceDelta delta
+        BalanceDelta delta,
+        bytes memory data
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:80]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 6]
```diff
@@ -78,7 +86,7 @@ interface IHooks {
     /// @param amount0 The amount of token0 being donated
     /// @param amount1 The amount of token1 being donated
     /// @return bytes4 The function selector for the hook
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:88]: Comment is missed for `data`

File: contracts/interfaces/IHooks.sol [hunk 7]
```diff
@@ -88,7 +96,7 @@ interface IHooks {
...
```
Comment (0xb10ckdev) [inline contracts/interfaces/IHooks.sol:98]: Comment is missed for `data`

**Review by snreynolds — DISMISSED** [review]
lgtm minus just the few comments left

File: contracts/PoolManager.sol [hunk 8]
```diff
@@ -144,11 +152,11 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC1155, IERC1155Rec
     }
 
     /// @inheritdoc IPoolManager
-    function lock(bytes calldata data) external override returns (bytes memory result) {
+    function lock(bytes calldata hookData) external override returns (bytes memory result) {
```
Comment (emmaguo13) [inline contracts/PoolManager.sol:155]: should this be called hookData? the data that gets passed into lock does not necessarily have to be for hooks

Comment (ewilz) [reply contracts/PoolManager.sol:155]: bahh no!! thanks. lol so many typos. everything's called data, it was hard not to grab some extra ones in there

**Review by emmaguo13 — APPROVED** [review]

**Pull request merged**